Hr update (#1)
nhs-r-community/git_training: guides/intro_to_git.md

> # Overview

Reviewer: I think this should be called 'Introduction to Git'

This page is intended as a starting point for R users who are new to using Git. It shows you the basic commands and the workflow to use Git on your own, in conjunction with using RStudio desktop or RStudio Cloud. 

To use Git as a team you should complete this section and the following section on [using git collaboratively](02_using-git-collaboratively.md).

## What is version control?

> Version control is the practice of tracking and managing any changes on project's code, files or folders over time. This allows you to observe a detailed history of the changes made and enables your team members to collaborate on the same project.

Reviewer: Not clear what Git is here, esp. considering the file is called intro_to_git

Reviewer: When you say 'projects, files and folders'- you might want to be clear that this is primarily for code, rather than tracking changes to word documents, and data files.

Author: I think we need a ticket to update the introduction_to_git over to RAP COP as well

 
In Git, each user has the entire repository (project's working directory) on their computer (offline), hence, they can work separately offline until they opt to push their updated version of the code to the remote (online) central repository. 

> Remote repositories are versions of your project that are hosted on an online Version Control System. In NHS Digital, we store that in GitLab, site for internal users (GitHub, site for the public).

Reviewer: Conscious some people at this training won't have a corporate GitLab/GitHub organisation and might need to create an account. Is it worth mentioning something like this- i.e. in order to host your code online, you will need to have an account with either, and we'd recommend github?

## Why should I care?

> Using version control is one of the fundamental skills needed by analysts to produce high-quality analytical outputs. Git is **the standard** for tracking code over time and is the way that NHSD has chosen. While Git has a steep initial learning curve, the payoff is huge and so we strongly recommend taking the time to learn this as a team.

Reviewer: Small thing but you might want to say NHS Digital- or be clear the first time above, maybe in brackets: NHS Digital (NHSD).

The benefits of using version control include:

* Understanding what happened in the past. E.g., what did we change last time we ran this code
* Restoring previous versions
* Tracking changes - avoid accidentally breaking code
* The ability to review someone's changes and to leave comments
* The ability to plan development work more effectively through being able to assign small, discrete changes
* Avoid code being hidden away on someone's machine
* The ability to set up an approval process for changes
* The ability to make changes without breaking anything - through running automated tests

> * The ability to try out experiments without the risk of breaking your main code

Reviewer: what about the ability to work on the same codebase collaboratively? (I appreciate this is outside the scope of this tutorial but is one of the biggest things some people miss about git when they are learning. It might be worth mentioning here with a caveat that people can learn more about that in the following section)

> # Glossary of terms

Reviewer: pet peeve- To conform to markdown best practices, header 1 # should only be used for the title.

Here's a list of commonly used Git terms, that will help you understand the technical jargon mentioned in the next sections:
* [Git terminology](https://www.javatpoint.com/git-terminology)

> # Common basic commands:

Reviewer: I find having this at the start is quite confusing to the user as they don't know what any of these commands are- perhaps we could have this at the end?

Below is a list of common commands for reference. We only list basic commands here. This is just to flag to you that these commands exist. 

> * [Git commands cheatsheet](https://training.github.com/downloads/github-git-cheat-sheet/)

Reviewer: This isn't really an item in the list- I think it should move to the above

* Commit staged files with a message: `git commit -m "<message>"`. Committing changes in this way captures a snapshot of the project's currently staged changes. Committed snapshots can be thought of as “safe” versions of a project

* Update your branch with the online/remote branch info: `git pull`. Pulls the latest changes from the repository, only affects the current active branch you’re in.

* Check the history of commits: `git log`. 
 
* Show unstaged changes between your index and working directory: `git diff`. Press `Q` to exit the diff log.

* Ignore files: `.gitignore file`. This file specifies untracked files that Git should ignore such as the sensitive information related to security or the data itself. Files already tracked by Git are not affected.

* Display a list of all branches in the repository: `git branch -a`.

* Delete a local branch: `git branch -D <your branch name>`.

### Undoing changes via Git commands

GitLab has a guide covering various cases on undoing any changes: https://docs.gitlab.com/ee/topics/git/numerous_undo_possibilities_in_git/

# Setup for Git Basics exercise

> ## Creating your access token on GitHub

Reviewer: Given this is an external audience- is it worth saying somewhere that these commands will only work if you have git installed?

remote: Authentication failed for 'repository address you're attempting to git clone'
```
This can mean a number of things: either your access token has expired or you have entered the wrong access token for your password (or username even). If your access token has expired or about to expire, you can check by going to Gitlab Profile -> Preferences -> Access Tokens then scroll all the way down and see if there’s an Active Access Token and check the expiration date. 
* These steps will help you fix your access to the repository:

1. To fix this we need to create a new access token (don’t assign a date this time, so it will never expire).

2. In the command terminal type ```git pull -v``` (v stands for verification). This should prompt a new Git credentials window.

* If you don't have an existing repository, and can't do step 2, then:

1. In the terminal type ```git clone <git repo's copied clone url>``` but with the new credentials included in the URL, so:
```
C:\My_documents>git clone https://<username>:<password>@<domain>/example.git 
```
So, in my case it would be:
```
C:\My_documents>git clone https://<username>:<MY_NEW_ACCESS_TOKEN>@<domain>/example.git 
```

> # Exercises 1 - 6: Accessing a Git repository, creating a branch, adding a new file, uploading your changes to GitHub

Reviewer: I'm a bit confused about why this is listed as 'exercises', but the bits below are 'general'.
Can't this just be 'getting started with git'?

Author: the intention is to have these as exercises and the General bit is extras e.g. how to do a merge request.

Author: but unsure now that you mention it. Feels clearer that these are exercises to be followed through?

Author: > Finally finished reviewing this file- a _lot_ of content! Needs a bit of refining and a possible re-structure. The most radical proposal discussed in these comments is the idea of potentially starting with git from the RStudio gui....

comment cut off?

**Getting started** - tasks 1 and 2 will be relevant only when accessing a repository for the **first time**.

### Video: Solution to exercises 1 - 6 (Using RStudio Desktop)
Here's the video solution to exercises 1-6: [Video exercises solution](https://vimeo.com/678172959)

> ## Using RStudio Desktop

Reviewer: Radical suggestion- we could teach people how to use git avoiding the command line?? (less of an entry barrier)- a bit like the instructions for RStudio cloud, but for the desktop app.

Author: need to show me! I'm intrigued

Reviewer: ![image](https://user-images.githubusercontent.com/28767009/194050213-086a68e4-01d1-4546-86b4-f14206c35ee4.png)
This kinda thing

Reviewer: It's actually pretty good! (I haven't tried fixing merge conflicts with it though, and I'm not sure whether you can git pull origin main)

Reviewer: But I would ignore for this PR as it's a bit too much

Author: this is a great suggestion and will have a look! Thanks

### 1. Creating your local Git repository folder

> i. Open a Command terminal. This can be any terminal, Anaconda Prompt or Git Bash etc. To access the terminal, click the Search option on the Windows toolbar and type "Anaconda Prompt" and open it. The file directory address that's displayed in the terminal will usually be the default top level C drive address or something similar.

Reviewer: perhaps include cmd in this list?

Author: great shout

iv. Back to the terminal (e.g. Command Prompt/Anaconda Prompt/Git Bash etc.), type `git branch -a` to view a list of all available branches existing in the repository. The branches that contain the `remote/` path are the branches online, on GitHub. Have a look at the list. You will notice your new branch is not there yet. To update your local master branch with all the lastest updates on Gitlab, type ```git pull```. Type `git branch -a` and you should see your branch available in the list now.

v. To start working on your branch and select, type in the terminal `git checkout <your branch name>`.

vi. Type ```git status``` and this way you can confirm that you are in a new branch. Git will display a message showing the current branch name and the current branch's information.

![](../images/image564.PNG)

_**Reminder:** any changes you make while in the main branch won't be uploaded to GitHub, as the main branch is protected and locked, so make sure you are in your own working branch! You can confirm this with step vi. above._

#### Branch naming strategy

> Following an agreed naming convention will make it much easier to keep the team's work organised. We follow use this approach:

Reviewer: ```suggestion
Following an agreed naming convention will make it much easier to keep the team's work organised. We follow this approach:
```

>     <project_id>_<initials>_<ticket_number>_<brief-description>

Reviewer: For an external audience I'd say take the project id and ticket number out

Author: nice one

Here the project_id refers to our project ID in confluence (DS218). The initials are from my name. The ticket number refers to the jira ticket number. 
E.g.:

    DS218_CQ_167_Add-field-definitions

Naming branches in this way helps to make it very clear what change should be happening in the branch. Sometimes code review will reveal that a branch that was intended to make one specific change in fact makes several changes. This should be avoided and the code review should reject the additional changes.

### 4. Add a new file

> By adding a new file to the project, you can avoid conflicts as each analyst will create and upload their own file.

Reviewer: I don't really understand what this means.

Author: ...that's a fair point.

There are many ways to add a new file to your branch. One way is to upload a file directly to your branch on GitHub. 

> The easiest way is to simply create the file (any type of file) in your working directory, while in your branch. You can also copy a file from another directory. The file you create or copy can be any type of file.

Reviewer: I think this is overly complicated 
```suggestion
There are several ways to add a new file to the repository. 

```

Author: agreed

> For this exercise, create a file while in RStudio, with simply right-clicking on the repository's folder and then selecting the Create New File option:

Reviewer: Doesn't this line need to be after the user has opened the project?

* To create a project from GitHub, select the "Your Workspace" option on the left hand side once in RStudio Cloud and select New Project on the far right of the top toolbar, in the Your Projects space.

![](../images/new_project_git.png)

* This will open a new dialogue that will ask for the Git Repository URL, this is the URL you copy, as described in [Step 2](01_intro_to_git.md#2-git-clone-copy-a-github-repository) of this guide. 
* You will be prompted to enter your username (GitHub username) and your password, which is the access token created from the [Create Access token step](01_intro_to_git.md#creating-your-access-token-on-github).
* Once the Git repository is deployed, you will notice the repository file structure on the right hand side panel.

![](../images/project_structure.png)

* Steps [3](01_intro_to_git.md#3-create-and-switch-to-a-new-branch), [4](01_intro_to_git.md#4-add-a-new-file) and [5](01_intro_to_git.md#5-commit-your-changes-follow-basic-git-command-workflow) will be the same, the only change being that instead of using your local terminal to input Git commands, you will be using the RStudio Cloud built-in terminal.

![](../images/terminal.png)

> # General: How to submit a pull request

Reviewer: Mentioned already above but I'm confused as to why this is 'General'.
It also doesn't follow on from the RStudio cloud section very well.

Author: hmm yeah the General bit is not 100% great, any suggestions for a different name? Or maybe move it to a different guide?

Reviewer: I made a separate comment for this ah but 'Pull Requests', or 'Code Review', or 'Using Git Collaboratively'?

**Pull request (PR)** is the application you submit on the GitHub repository, to announce to other collaborators working on the repository that you have a new change ready to be merged/part of the main build, the main branch of the repository. Then, either your collaborators or someone you assign will review the change and the branch you have been working on to determine whether the change is ready to be merged with the main branch. The equivalent term for GitLab users is Merge Request (MR).

To submit a pull request:

> 1. Simply head over to GitHub and to the respective repository

Reviewer: Love the optimism but not sure it will feel very simple!
```suggestion
1. Navigate to your repository on GitHub
```

2. Select Pull requests on the top bar menu. 

3. Either select New Pull Request (big green button) on the right hand side or select Compare & pull request if you recently pushed new changes to the branch.

4. Base ref will default to the main branch, head ref will be the branch you wish to merge with the main branch. You can change any of these options, there will be occasions you won't always aim to merge every branch to the main branch.

5. Click Create Pull Request.

> 6. Fill out any information related to the Pull request, title, description, any comments. On the rigth hand side you can assign yourself as the Assignee, and assign someone in your team to be the Reviewer. Then, select Create Pull Request.

Reviewer: ```suggestion
6. Give the Pull Request a title that summarises the changes you are proposing in a few words. Add any more detail that might help reviewers understand your changes in the 'description' section. On the right hand side you can assign yourself as the Assignee, and assign someone in your team to be the Reviewer. Then, select Create Pull Request.
```

Author: nice one

# General: How to accept a pull request

> Once you are assigned the pull request's reviewer, you should receive an email notification from GitHub with a link to the pull request's webpage. You can also view any active pull requests by clicking on Pull Requests on the top bar menu.

Reviewer: ```suggestion
If you are assigned as the Reviewer of a Pull Request, you should receive an email notification from GitHub with a link to the pull request's webpage. You can also view any active pull requests by clicking on Pull Requests on the top bar menu.
```

> Once you are assigned the pull request's reviewer, you should receive an email notification from GitHub with a link to the pull request's webpage. You can also view any active pull requests by clicking on Pull Requests on the top bar menu.

Reviewer: Is it worth explaining the general process of reviewing, rather than listing the different views? I feel like if you are new to this, you don't necessarily need a description of what everything is, but you do want to know what to look at (e.g. Files changed)- I rarely look at commits or Checks

Author: ah I see what you mean now. Yeah, fair point, I think it might be worth splitting this section out from the guide into its own guide basically

![](../images/pull_requests.png)

On the Pull request page, there are four tabs: Conversation, Commits, Checks and Files changed:

### Conversation

> **Conversation** is the main pull request page. It contains the pull request's description, timeline and comments left by the assignee or reviewer. To merge a pull request you have 3 presented options:

Reviewer: be consistent with capitalisation of Pull Request (not sure which way to go)

Author: ahhhhh

* Merge pull request: this will keep all commits history of the source branch and will be added to the target branch. Creates a symbolic commit.

> * Squash and merge: ticking this will merge the history of commits for the source branch into one commit. This can be useful when the pull request is not a major update and you don't want your repository's history of commits to contain commits with messages such as "fixed typo", "replaced incorrectly placed image". Creates a symbolic commit.

Reviewer: Not sure 'creates a symbolic commit' is needed here as it's not clear what this is.

> * Rebase and merge: adds all source branch commits in front of the target branch's commits. Does not create a symbolic commit.

Reviewer: Do we need to explain what this is here? We don't explain _why_ you would want to do this so I don't see how we are adding info here.

Author: sorry, your comment has confused me

Reviewer: So- why would you want to rebase and merge, instead of just 'merge'?

Author: oh, I'm just going through the various options available, as part of the guide but not actually explaining every option during the workshop except for squash and merge.

### Commits

**Commits** tab displays a history of the pull request's commits. So for example, if the reviewer decides that minor adjustments are required and the assignee pushes these changes to Gitlab, then these commits will be captured in this tab.

### Files changes

**Files changed** tab, here you can see all the new changes in the pull request's source branch. You can also opt for a side-by-side comparison, and see line-by-line how the master version and new branch version compare. You can also leave comments on each file's line, which will also be recorded in the Conversation tab's timeline.

### Checks

**Checks** is part of the GitHub Actions functionality, which enchances your code review processes with apps and automated reviewing checks. This is inactive on default.

If you are happy to merge the two branches, then select the type of Pull Request you wish to apply and the pull request should be completed in a few seconds.

> # General: How to update your local main branch to the latest remote main version on GitHub/GitLab

Reviewer: Think we are missing a why/when you would want to do this here.

Author: nice one

> 1. Open the Anaconda Prompt terminal or Git Bash or any other terminal you wish to use. This can be accessed by typing, for example Anaconda Prompt, in the search bar, after opening the search icon on your Windows toolbar.

Reviewer: Do we need to repeat this line about terminal usage everywhere? If we do, we should at least be consistent.

2. Type ```cd <paste your work directory address>``` in the terminal to switch to your working directory. This is where the Git repository is cloned/copied/downloaded to, on your local machine (or RDS environment or any environment you're using). This will cause the default terminal address to be directed at the new directory. (see [Git Clone instructions above](01_intro-to-git.md#3-git-clone-copydownload-a-gitlab-repository)  for more info)

3. Type ```git status``` in the terminal to ensure that ```ii``` happened successfully and you are on the default main branch. You should see this message: 

![](../images/image3.jpg)

4. Type ```git pull``` to update your local main branch with the latest version of the remote/online GitHub main branch. This might take a few seconds.

(Optional) 5. If you're in a different branch and not the main branch, then type in the terminal ```git checkout main``` and then ```git status``` to confirm you are on the main branch.

> # General: the .gitignore file

Reviewer: If we have .gitignore here (v important!), should we also have README.md? I'm conscious R users may be very familiar with markdown in general but may have never seen a README file.

> .gitignore is a text file that contains file extensions and directories' paths that we wish Git to ignore. For example, we have created a repository on GitHub that should never contain any sensitive data. To ensure this, in the repository's .gitignore file we will include ```.csv```, `.xlsx` or any other file format that can contain data.

Reviewer: We need to add a .gitignore (see #3 )- but also need to add .csv and .xlsx.
Worth saying again that I'm not sure we've mentioned the data upload risk yet.

Author: I've merged your branch (containing .gitignore), might be worth having a call to go this the comments I replied to above

Reviewer: Can you pull main into this branch, and then we can see the gitignore.

Author: done!

> GitHub has a .gitignore [template](https://github.com/github/gitignore/blob/main/R.gitignore) available to analysts and developers to use in their projects. Notice how in the template, line 22 contain RStudio project settings, which are configuration files created by RStudio when you first open your repository as an RStudio project. Sometimes these configuration files can contain information you don't want to publish on GitHub/GitLab, thus by including them in the .gitignore file you are ensuring that these files will never leave your local machine.

Reviewer: These files are included in the .gitignore in #3  PR- might be worth referencing.

> # External links

Reviewer: Could we add the git cheatsheet here?
